langchain4j/src/main/java/dev/langchain4j/data/document/loader/AbstractS3Loader.java

```diff
@@ -0,0 +1,171 @@
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.S3ClientBuilder;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static dev.langchain4j.internal.Utils.isNotNullOrBlank;
+import static dev.langchain4j.internal.Utils.isNullOrBlank;
+
+public abstract class AbstractS3Loader<T> {
+    protected final String bucket;
+    protected final String region;
+    protected final String endpointUrl;
+    protected final String profile;
+
+    protected final boolean forcePathStyle;
+    protected final AwsCredentials awsCredentials;
+
+    protected AbstractS3Loader(Builder builder) {
+        this.bucket = builder.bucket;
+        this.region = builder.region;
+        this.endpointUrl = builder.endpointUrl;
+        this.profile = builder.profile;
+        this.forcePathStyle = builder.forcePathStyle;
+        this.awsCredentials = builder.awsCredentials;
+    }
+
+    /**
+     * Initiates the loading process by configuring the AWS credentials and S3 client,
+     * then loads either a document or a list of documents.
+     *
+     * @return A generic object of type T, which could be a Document or a list of Documents
+     * @throws RuntimeException if there are issues with AWS credentials or S3 client configuration
+     */
+    public T load() {
+        AwsCredentialsProvider awsCredentialsProvider = configureCredentialsProvider();
+        S3Client s3Client = configureS3Client(awsCredentialsProvider);
+        return load(s3Client);
+    }
+
+    private static AwsSessionCredentials toAwsSessionCredentials(AwsCredentials awsCredentials) {
+        return AwsSessionCredentials.create(awsCredentials.getAccessKeyId(), awsCredentials.getSecretAccessKey(), awsCredentials.getSessionToken());
+    }
+
+    private static software.amazon.awssdk.auth.credentials.AwsCredentials toAwsCredentials(AwsCredentials awsCredentials) {
+        return AwsBasicCredentials.create(awsCredentials.getAccessKeyId(), awsCredentials.getSecretAccessKey());
+    }
+
+    protected abstract T load(S3Client s3Client);
+
+    private AwsCredentialsProvider configureCredentialsProvider() {
+        AwsCredentialsProvider provider = DefaultCredentialsProvider.builder().build();
+
+        if (awsCredentials != null) {
+            if (awsCredentials.hasAccessKeyIdAndSecretKey()) {
+                provider = StaticCredentialsProvider.create(toAwsCredentials(awsCredentials));
+            } else if (awsCredentials.hasAllCredentials()) {
+                provider = StaticCredentialsProvider.create(toAwsSessionCredentials(awsCredentials));
```

Comment:
it seems that this code block is not reachable. if all 3 arguments are present and you always check `hasAccessKeyIdAndSecretKey()` first, it will always execute only first if block

Reply:
nice catch. updated the code.